main/src/cgeo/geocaching/downloader/DownloaderUtils.java

```diff
@@ -0,0 +1,280 @@
+import android.widget.CheckBox;
+import android.widget.TextView;
+import android.widget.Toast;
+import static android.app.Activity.RESULT_OK;
+import static android.content.Context.DOWNLOAD_SERVICE;
+
+import androidx.annotation.Nullable;
+import androidx.annotation.StringRes;
+
+import java.lang.ref.WeakReference;
+import java.util.ArrayList;
+import java.util.List;
+
+import okhttp3.Response;
+import org.apache.commons.lang3.StringUtils;
+
+public class DownloaderUtils {
+
+    public static final int REQUEST_CODE = 47131;
+    public static final String RESULT_CHOSEN_URL = "chosenUrl";
+    public static final String RESULT_SIZE_INFO = "sizeInfo";
+    public static final String RESULT_DATE = "dateInfo";
+    public static final String RESULT_TYPEID = "typeId";
+
+    private DownloaderUtils() {
+        // utility class
+    }
+
+    public static boolean onOptionsItemSelected(final Activity activity, final int id) {
+        if (id == R.id.menu_download_offlinemap) {
+            activity.startActivityForResult(new Intent(activity, MapDownloadSelectorActivity.class), REQUEST_CODE);
+            return true;
+        }
+        return false;
+    }
+
+    public static boolean onActivityResult(final Activity activity, final int requestCode, final int resultCode, final Intent data) {
+        if (requestCode == REQUEST_CODE && resultCode == RESULT_OK) {
+            // trigger download manager for downloading the requested file
+            final Uri uri = data.getParcelableExtra(RESULT_CHOSEN_URL);
+            final String sizeInfo = data.getStringExtra(RESULT_SIZE_INFO);
+            final long date = data.getLongExtra(RESULT_DATE, 0);
+            final int type = data.getIntExtra(RESULT_TYPEID, Download.DownloadType.DEFAULT);
+            if (null != uri) {
+                triggerDownload(activity, R.string.downloadmap_title, type, uri, "", sizeInfo, date, null);
+            }
+            return true;
+        }
+        return false;
+    }
+
+    public static void triggerDownload(final Activity activity, @StringRes final int title, final int type, final Uri uri, final String additionalInfo, final String sizeInfo, final long date, final Runnable callback) {
+        String temp = uri.getLastPathSegment();
+        if (null == temp) {
+            temp = "default.map";
+        }
+        final String filename = temp;
+
+        final AlertDialog.Builder builder = Dialogs.newBuilder(activity);
+        builder.setTitle(title);
+        final View layout = View.inflate(activity, R.layout.mapdownloader_confirmation, null);
+        builder.setView(layout);
+        final TextView downloadInfo = layout.findViewById(R.id.download_info);
+        downloadInfo.setText(String.format(activity.getString(R.string.download_confirmation), StringUtils.isNotBlank(additionalInfo) ? additionalInfo + "\n\n" : "", filename, "\n\n" + activity.getString(R.string.download_warning) + (StringUtils.isNotBlank(sizeInfo) ? "\n\n" + sizeInfo : "")));
```

Comment:
I detect that this code is problematic. According to the [Bad practice (BAD_PRACTICE)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#bad-practice-bad-practice), [FS: Format string should use %n rather than n (VA_FORMAT_STRING_USES_NEWLINE)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#fs-format-string-should-use-n-rather-than-n-va-format-string-uses-newline).
This format string includes a newline character (\n). In format strings, it is generally  preferable to use %n, which will produce the platform-specific line separator.

Reply:
Thanks for pointing me to that info. Alas, it seems to not apply to Android apps and/or not in the context of dialogs: I've tried switching the string to using `%n`, but the placeholders get printed as literal `%n` in the dialog, so I reverted them to `\n`.

Reply:
noticed. thanks a lot!